geopandas/geodataframe.py

```diff
@@ -1425,6 +1419,27 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
 
     @property
     def _constructor(self):
+        geometry_default = self._geometry_column_name
+        crs_default = self.crs
+
+        def _geodataframe_constructor_with_fallback(
+            data=None, index=None, crs=crs_default, geometry=geometry_default, **kwargs
+        ):
+            df = pd.DataFrame(data, index=index, **kwargs)
+            if geometry in df.columns:
+                try:
```

Comment:
This had to be changed to a try/ except block because sometimes an ensure_geometry call is needed - noticed from test_plotting.py::test_geometry_geometry which calls `gdf.loc[3] = [None]` where 3 is a new row not currently in `gdf`, and I've added `test_setitem_new_index` below to cover this explicitly (makes me wonder how much other behaviour this could break).